scripts/proto/lbann_modules.py

```diff
@@ -0,0 +1,221 @@
+        return (obj,)
+
+class Module:
+    """Base class for neural network modules.
+
+    A module is a pattern of operations that may be applied to a set
+    of input layers, obtaining a set of output layers.
+
+    """
+
+    def __init__(self):
+        pass
+
+    def forward(self, input):
+        """Apply module pattern to `input`.
+
+        `input` is a `Layer` or a sequence of `Layer`s. The module
+        pattern is added to the layer graph and the output layer(s)
+        are returned.
+
+        """
+        # Should be overridden in all sub-classes
+        raise NotImplementedError
+
+    def __call__(self, input):
+        """Apply module mattern to `input`.
+
+        Syntatic sugar around `forward` function.
+
+        """
+        return self.forward(input)
+
+class FullyConnectedModule(Module):
+    """Basic block for fully-connected neural networks.
+
+    Applies a dense linearity and a nonlinear activation function.
+
+    """
+
+    global_count = 0  # Static counter, used for default names
+
+    def __init__(self, size, bias=True, weights=[], activation=None,
+                 name='', data_layout='data_parallel'):
+        """Initialize fully-connected module.
+
+        Args:
+            size (int): Size of output tensor.
+            activation (type): Layer class for activation function.
+            bias (bool): Whether to apply bias after linearity.
+            weights (`Weights` or iterator of `Weights`): Weights in
+                fully-connected layer. There are at most two: the matrix
+                and the bias. If weights are not provided, LBANN will
+                initialize them with default settings.
+            name (str): Default name is in the form 'fcmodule<index>'.
+            data_layout (str): Data layout.
+
+        """
+        super().__init__()
+        FullyConnectedModule.global_count += 1
+        self.instance = 0
+        self.size = size
+        self.bias = bias
+        self.name = (name
+                     if name
+                     else 'fcmodule{0}'.format(FullyConnectedModule.global_count))
+        self.data_layout = data_layout
+
+        # Initialize weights
+        # Note: If weights are not provided, matrix weights are
+        # initialized with He normal scheme and bias weights are
+        # initialized with zeros.
+        self.weights = list(_make_iterable(weights))
+        if len(self.weights) > 2:
+            raise ValueError('`FullyConnectedModule` has '
+                             'at most two weights, '
+                             'but got {0}'.format(len(self.weights)))
+        if len(self.weights) == 0:
+            self.weights.append(
+                lp.Weights(initializer=lp.HeNormalInitializer()))
+        if len(self.weights) == 1:
+            self.weights.append(
+                lp.Weights(initializer=lp.ConstantInitializer(value=0.0)))
+
+        # Initialize activation layer
+        self.activation = None
+        if activation:
```

Comment:
This prevents using modules as an action (if, for whatever reason, that makes sense). Not sure if we care.

Reply:
Thinking about it, a module is actually more natural here. We are basically mimicking a module that applies an activation function. Perhaps we could implement a module that wraps around a single layer? If the user provides a layer, we can construct the corresponding module. Or maybe it would be messier than it's worth.

Reply:
That might be more complex than it's worth. Alternately, we could make layers a subclass of modules. This currently seems like an edge-case we can address later though.